src/napi/js_native_ext_api_v8.cpp

```diff
@@ -0,0 +1,273 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ----------------------------------------------------------------------------
+// Some code is copied from Node.js project to compile V8 NAPI code
+// without major changes.
+// ----------------------------------------------------------------------------
+// Original Node.js copyright:
+//
+// Copyright Joyent, Inc. and other Node contributors.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to permit
+// persons to whom the Software is furnished to do so, subject to the
+// following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
+// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+// USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+#include "env-inl.h"
+
+#include "V8JsiRuntime_impl.h"
+#include "js_native_api_v8.h"
+#include "js_native_ext_api.h"
+#include "public/ScriptStore.h"
+
+static struct napi_env_scope__ {
```

Comment:
>napi_env_scope__ [](start = 14, length = 16)

I like our current approach of creating these scope objects in stack better than this..

Reply:
Could you please explain what do you mean by that?
The C-based API cannot have stack-based objects because C does not have classes and destructors.
These C-based must not be used directly in C++ for exception-safety, they must be wrapped up into C++ classes that we put on the call stack which will automatically open/close the scope.